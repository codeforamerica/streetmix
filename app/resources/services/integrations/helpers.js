const { User } = require('../../../db/models')

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
const findUser = async function (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
const addOrUpdateByProviderName = function (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
const syncAccountStatus = async function (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

module.exports = { findUser, addOrUpdateByProviderName, syncAccountStatus }
