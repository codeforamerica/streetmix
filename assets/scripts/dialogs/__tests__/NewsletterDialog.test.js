/* eslint-env jest */
import React from 'react'
import { cleanup } from '@testing-library/react'
import { renderWithReduxAndIntl } from '../../../../test/helpers/render'
import NewsletterDialog from '../NewsletterDialog'

// Mock Mailchimp HTML snippet
jest.mock('../Newsletter/mailchimp.html', () => '<div>foo</div>')

describe('NewsletterDialog', () => {
  afterEach(cleanup)

  it('renders snapshot', () => {
    const wrapper = renderWithReduxAndIntl(<NewsletterDialog />)
    expect(wrapper.asFragment()).toMatchSnapshot()
  })
})
