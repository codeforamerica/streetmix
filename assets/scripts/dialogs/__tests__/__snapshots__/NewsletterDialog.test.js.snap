// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsletterDialog renders snapshot 1`] = `
<DocumentFragment>
  <div
    class="dialog-box-container"
  >
    <div
      class="dialog-box-backdrop"
    />
    <div
      class="dialog-box"
      role="dialog"
    >
      <button
        class="close"
        title="Dismiss"
      >
        <svg
          class="svg-inline--fa fa-times"
        />
      </button>
      <div
        class="newsletter-dialog"
      >
        <header>
          <h1>
            <span>
              Subscribe to our newsletter
            </span>
          </h1>
        </header>
        <div
          class="dialog-content"
        >
          <div>
            <div>
              foo
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
